### PR TITLE
biber 2.16 (new formula)

### PR DESCRIPTION
**Pull request by max-ae**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----
This is a followup to #87190 and aims to create a formula for biber.

The resources build fine and the `Build` script does not complain about missing dependencies, also the `Build install` step runs fine, but I still can't produce a working binary...

Maybe someone (more familiar with perl) can help?

Also, a test is still missing since the formula does not work yet.

### CODE REVIEW

**Comment by max-ae** [comment]
@danielnachun Did you have time to look at a local build yet? 😃

**Comment by carlocab** [comment]
Btw, I didn't mean to make this as ready for review. I opened this in the GitHub app on my phone when I got a notification for it and accidentally pressed the "Mark Ready for Review" button while scrolling. 😅

**Comment by danielnachun** [comment]
@max-ae I have indeed been working on this and made progress, though it's not totally working yet.  I put what I have so far into this gist: https://gist.github.com/danielnachun/8ba200b5350b4c4a898ef9d2c590ae57.  There are a few changes:

1) I reviewed all the Perl resources and removed any modules which are provided by brewed Perl.  We shouldn't normally need to install those unless a specific version is needed, in part because they will create files that conflict with those in brewed Perl like we saw here.

2) I also removed the `Test`/`CPAN`/`Dist` modules as we don't run tests with `make check` or try to look up any packages up on CPAN in the build process.  There were a small number of other modules which did not seem needed anymore which I suspect were dependencies of the modules I removed.

3) Conversely, I did identify a handful of modules that seemed to be missing and added them.  Overall we are still installing less modules now that before.

4) I tweaked the code for installing the Perl modules so that we don't have to hardcode which modules use `Build.PL` vs `Makefile.PL`but instead detect this automatically.  That will save some effort if in the future new modules have to be added so that we don't have to figure out which build system they use.

As for what currently does not work, while it builds fine on Linux I am hitting a build error for `Net::SSLeay` in macOS.  I encountered what I think was the same error when trying to install this for `texlive`, and I thought that adding `ENV["OPENSSL_PREFIX"] = Formula["openssl@1.1"].opt_prefix` would fix it, but on my Intel Monterey machine it does seem to help now.  I set up the `texlive` formula on Big Sur so it's possible this is a new issue in Monterey.

Once this works I'd like to revisit trying to use the system Perl on the macOS install if possible.  The module that was failing before was one which is provided by both brewed and macOS system Perl, so I was able to remove it.  That may mean we can use the system Perl on macOS but it will require figuring out which modules are provided by the system Perl but not brewed Perl.

@max-ae feel free to update your PR with the formula I linked above so we can see where it is failing for different OSes.   I added the `CI-no-fail-fast` label so that all runners will try to finish even if one fails.  This will let us know which platforms need more work and what exactly is failing.

**Comment by danielnachun** [comment]
I did more testing and the system Perl on macOS seems to be too old:
```
    !  perl (5.30.3) is installed, but we need version >= 5.32.0
```

**Comment by max-ae** [comment]
Build now fails with this:

```
*** Found OpenSSL-1.1.1l installed in /usr/local/opt/openssl@1.1
*** Be sure to use the same compiler and options to compile your OpenSSL, perl,
    and Net::SSLeay. Mixing and matching compilers is not supported.
Checking if your kit is complete...
Looks good
WARNING: /usr/local/opt/perl/bin/perl is loading libcrypto in an unsafe way
```

I can reproduce this locally on Intel Monterey, but have no idea how to fix this

**Comment by carlocab** [comment]
> Build now fails with this:
> 
> ```
> *** Found OpenSSL-1.1.1l installed in /usr/local/opt/openssl@1.1
> *** Be sure to use the same compiler and options to compile your OpenSSL, perl,
>     and Net::SSLeay. Mixing and matching compilers is not supported.
> Checking if your kit is complete...
> Looks good
> WARNING: /usr/local/opt/perl/bin/perl is loading libcrypto in an unsafe way
> ```
> 
> I can reproduce this locally on Intel Monterey, but have no idea how to fix this

I think that comes from this change in behaviour in `dyld`: https://openradar.appspot.com/FB9725981

**Comment by max-ae** [comment]
@danielnachun The only failure now is a linkage failure on Linux, but I don't know how to resolve this.
There are some remarks in the upstream buildscript for Linux about `libbtparse.so`, but I don't know what to make of it since I don't know how shared libraries work on Linux.
https://github.com/plk/biber/blob/34a6e574e441ea0596577d698c24544016e23fa6/dist/linux_x86_64/build.sh#L17-L19
https://github.com/plk/biber/blob/34a6e574e441ea0596577d698c24544016e23fa6/dist/linux_x86_64/build.sh#L73

**Comment by danielnachun** [comment]
Nice work on the patch!  The `libbtparse.so` issue is slightly complex.  The root cause is that `libbtparse.so` is installed in `libexec/lib/`, which is not added by default to the RPATH for Linux, so `BibTeX.so` can't find the library.  

There are two solutions to this: 
1) Symlink `libbtparse.so` to `lib` so that it will be in the default `RPATH` we add for formulae (I've tested this manually and it does fix the problem).  The downside to doing this is that if any other package depends on `Text::BibTeX`, it would have a file name conflict.  
2) Figure out how to add `libexec/lib` to the RPATH on Linux.  So far I've tried the usual `LDFLAGS` approach and it's ignored.  To my disappointment, it also seems the package hard codes `extra_linker_args` to `-Lbtparse/src -lbtparse`.  We should be able to use an `inreplace` to insert `-Wl,-rpath,#{libexec}/lib` in front of `-lbtparse`.  I'm looking into that now.

File: Formula/biber.rb [hunk 1]
```diff
@@ -0,0 +1,567 @@
...
+
+    resources.each do |r|
+      next if testresources.include?(r.name)
+
+      r.stage do
+        if File.exist? "Makefile.PL"
+          system "perl", "Makefile.PL", "INSTALL_BASE=#{libexec}"
+          system "make"
+          system "make", "install"
+        else
```
Comment (danielnachun) [inline Formula/biber.rb:528]: Try to add this block here before call `perl Build.PL` to fix the test failure on Linux:
```
          if r.name == "Text::BibTeX" && OS.linux?
            inreplace "inc/MyBuilder.pm",
                      "-lbtparse",
                      "-Wl,-rpath,#{libexec}/lib -lbtparse"
          end
```

**Review by danielnachun — APPROVED** [review]
Really great work on this!  I'll leave a day or so for other maintainers to review and then I'll publish the bottles.

**Pull request closed without merging**